### PR TITLE
Refactor Tensor Mechanics Guarantees

### LINKED ISSUE #8881: "Guarantees" - MaterialProperty meta data
### Description of the enhancement or error report

I would be nice if MOOSE provided a way to decorate material properties with a specific type of meta data: guarantees that the material property satisfies. Guarantees could be given during material construction. And they could be enforced hard when getting the property (with MOOSE erroring out if not all materials that are supplying the property on the relevant subdomains make a required guarantee), or they could be queries softly and used to inform a cod path choice.

**Example:** A `RankFourTensor` (elasticity tensor) could be guaranteed to be isotropic.
* Certain calculations are dramatically _faster_ on isotropic tensors (divide by scalar) than on anisotropic tensors (invert a rank four tensor). 
* Some of our models yield _incorrect_ results when used with anisotropic tensors (finite strain models neglect to rotate the elasticity tensor into the current deformed configuration).

### Rationale for the enhancement or information for reproducing the error

Guarantees on material properties would allow us to
* implement incomplete models (like finite strain currently is) and properly warn our users when they are used improperly. 
* implement accelerated code paths (skip rotations and inverts on isotropic tensors etc.)

### Identified impact
New optional capability.

### PR DESCRIPTION
**Pull request by dschwen**
This adds one enum class and two new classes that can be inherited from in addition to the MOOSE base classes

* `Guarantee` is an `enum class` that contains the set of valid guarantees. This can be expanded as necessary.
* `GuaranteeProvider` can be added to Material classes and provides the `void issueGuarantee(MaterialPropertyName, Guarantee)` function that add guarantee metadata to a material property declared by the class.
* `GuaranteeConsumer` can be added to Materials and kernels and provides the `bool hasGuaranteedMaterialProperty(MaterialPropertyName, Guarantee)` method that checks if all materials on each block that provide a given material property issue the required guarantee.

Closes #8881 (because this as close as we will ever get this)

### CODE REVIEW

**Comment by sapitts** [comment]
I vaguely remember _someone_ telling @bwspenc that using Atom would prevent these sorts of clang formatting errors....
But in seriousness, this line with the error message, which has obviously been formatted via clang format once, caught me on a merge request too; is there a way to prevent clang format from complaining about changes which it made itself on a previous commit?

**Comment by dschwen** [comment]
Bison PR is standing by.

File: modules/tensor_mechanics/src/utils/GuaranteeConsumer.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/utils/GuaranteeConsumer.C]: This has general applicability beyond the elasticity tensor, so I'd change "elasticity tensor" to "material property"

File: modules/tensor_mechanics/src/utils/GuaranteeConsumer.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/utils/GuaranteeConsumer.C]: Per the above comment, 'elastic_mat' probably isn't a great variable name here.

File: modules/tensor_mechanics/include/utils/Guarantee.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/utils/Guarantee.h]: Would it be better to name this something like 'PropertyGuarantee' just to be a little more descriptive about what it does?  I don't want to make everything have huge names, and I'm not that set on renaming it, so I'll leave that up to you.

**Review by bwspenc — CHANGES_REQUESTED** [review]
This looks good.  Just a few very minor things to address.

**Comment by dschwen** [comment]
@sapitts are you waiting for #9484 to use the new guarantees system to implement this?

**Review by bwspenc — APPROVED** [review]

**Comment by bwspenc** [comment]
This breaks BISON and MASTODON, but we have patches ready.

**Pull request merged**
